File: Stack/150. Evaluate Reverse Polish Notation /EvaluateReversePolishNotation.cpp
//
//  EvaluateReversePolishNotation.cpp
//  LeetCode-main
//
//  Created by jackma on 2022/6/5.
//  Copyright © 2022 JackMa. All rights reserved.
//

#include "EvaluateReversePolishNotation.hpp"
#include <stack>

/**
 Evaluate the value of an arithmetic expression in Reverse Polish Notation.
 
 Valid operators are +, -, *, and /. Each operand may be an integer or another expression.
 
 Input: tokens = ["2","1","+","3","*"]
 Output: 9
 Explanation: ((2 + 1) * 3) = 9
 
 Input: tokens = ["10","6","9","3","+","-11","*","/","*","17","+","5","+"]
 Output: 22
 Explanation: ((10 * (6 / ((9 + 3) * -11))) + 17) + 5
 = ((10 * (6 / (12 * -11))) + 17) + 5
 = ((10 * (6 / -132)) + 17) + 5
 = ((10 * 0) + 17) + 5
 = (0 + 17) + 5
 = 17 + 5
 */

void testEvaluateReversePolishNotation() {
    vector<string> tokens = {"4","13","5","/","+"};
    cout << Solution().evalRPN(tokens) << endl;
}

int Solution::evalRPN(vector<string> &tokens) {
    stack<int> rpnStack;
    for (string token : tokens) {
        switch (token[token.length()-1]) {
            case '+': {
                int num1 = rpnStack.top();
                rpnStack.pop();
                int num2 = rpnStack.top();
                rpnStack.pop();
                rpnStack.push(num2+num1);
            }
                break;
            case '-': {
                int num1 = rpnStack.top();
                rpnStack.pop();
                int num2 = rpnStack.top();
                rpnStack.pop();
                rpnStack.push(num2-num1);
            }
                break;
            case '*': {
                int num1 = rpnStack.top();
                rpnStack.pop();
                int num2 = rpnStack.top();
                rpnStack.pop();
                rpnStack.push(num2*num1);
            }
                break;
            case '/': {
                int num1 = rpnStack.top();
                rpnStack.pop();
                int num2 = rpnStack.top();
                rpnStack.pop();
                rpnStack.push(num2/num1);
            }
                break;
            default: {
                int num = stoi(token);
                rpnStack.push(num);
            }
                break;
        }
    }
    
    return rpnStack.top();
}

File: Stack/150. Evaluate Reverse Polish Notation /EvaluateReversePolishNotation.swift
//
//  EvaluateReversePolishNotation.swift
//  LeetCode-Swift
//
//  Created by jackma on 2023/4/14.
//  Copyright © 2023 JackMa. All rights reserved.
//

import Foundation

/**
 https://leetcode.com/problems/evaluate-reverse-polish-notation
 
 You are given an array of strings tokens that represents an arithmetic expression in a Reverse Polish Notation.

 Evaluate the expression. Return an integer that represents the value of the expression.

 Note that:

 The valid operators are '+', '-', '*', and '/'.
 Each operand may be an integer or another expression.
 The division between two integers always truncates toward zero.
 There will not be any division by zero.
 The input represents a valid arithmetic expression in a reverse polish notation.
 The answer and all the intermediate calculations can be represented in a 32-bit integer.
 */

extension Solution {
    func EvaluateReversePolishNotation(_ tokens: [String]) -> Int {
        var stack = Array<Int>()
        
        for token in tokens {
            switch token.last {
            case "+": do {
                let num1 = stack.last!
                stack.removeLast()
                let num2 = stack.last!
                stack.removeLast()
                stack.append(num1+num2)
            }
                break
            case "-": do {
                let num1 = stack.last!
                stack.removeLast()
                let num2 = stack.last!
                stack.removeLast()
                stack.append(num2-num1)
            }
                break
            case "*": do {
                let num1 = stack.last!
                stack.removeLast()
                let num2 = stack.last!
                stack.removeLast()
                stack.append(num2*num1)
            }
                break
            case "/": do {
                let num1 = stack.last!
                stack.removeLast()
                let num2 = stack.last!
                stack.removeLast()
                stack.append(num2/num1)
            }
                break
            default:
                let num = Int(token)!
                stack.append(num)
            }
        }
        
        return stack.last!
    }
}
